src/c4_client/mod.rs

```diff
@@ -0,0 +1,62 @@
+//! Until now we have focused primarily on the blockchain as a data structure. We've created instances of the
+//! data structure, practiced validating it, and deciding on a canonical branch when forks occur and the
+//! data structure becomes more like a tree than a list.
+//!
+//! Even as we learned how to abstract out the common elements of the blockchain such as the consensus rules,
+//! and state machine logic, we still remained focused on the data structure itself.
+//!
+//! In this final chapter, we will shift our focus toward a blockchain client. A client is a piece of software
+//! that follows a blockchain in real-time. It imports blocks, follows forks, queues transactions, and even authors blocks.
+//! Throughout this chapter, we will use the state machine and consensus abstractions that we developed in the
+//! previous two chapters.
+
+// TODO Exercise for later: Client does a hard fork at a particular block height. The fork logic is to change runtimes.
+
+use crate::{
+    c1_state_machine::StateMachine,
+    c3_consensus::{Consensus, Header},
+};
+use p1_data_structure::Block;
+pub use p3_fork_choice::ForkChoice;
+use std::{
+    collections::{HashMap, HashSet},
+    marker::PhantomData,
+};
+
+mod p1_data_structure;
+mod p2_importing_blocks;
+mod p3_fork_choice;
+mod p4_transaction_pool;
+mod p5_authoring_blocks;
+mod p6_finality;
+
+type Hash = u64;
+
+/// A client represents one view of an evolving blockchain network. It knows of blocks,
+/// forks, state, and it also pools transactions waiting to be included in upcoming blocks.
+/// It can import new blocks, author its own blocks.
+///
+/// The client that we are writing is very reusable and is generic over multiple state machines
+/// and consensus systems. These are represented as generic parameters.
+///
+/// As you work through the sections in this chapter, you will add features to the client
+/// by implementing more and more traits on it. You may also need to revisit your struct definition
+/// to add more fields.
+pub struct FullClient<C, SM, FC>
+where
+    C: Consensus,
+    SM: StateMachine,
+    FC: ForkChoice<C>,
+{
+    // You are free to add fields here. Please document them as you add them.
+    _ph_data: PhantomData<(C, SM, FC)>,
+
+    //TODO Probably don't want to include all of these right off the bat.
+    // For now it is more of a brain dump.
+    transaction_pool: Vec<SM::Transition>,
+    // block_database: HashMap<Hash, Block>,
+    // state_database: HashMap<Hash, State>,
+    leaves: HashSet<Hash>,
```

Comment:
What are the leaves in this case?

Reply:
To answer the question: The leaves are all of the tips of the blockchain tree. A block that doesn't yet have any children is a leaf. Only leaves are candidates to be the longest chain.

But also, I removed this from the starter code. This was part of my braindump, and I think it is cleaner now. And actually, I think you only need to explicitly track the leaves in some fork choice rules, so it definitely shouldn't go directly on the client struct.